File: leet_0054_spiral_matrix.ts
function spiralOrder(g: number[][]): number[] {
    let res: number[] = []
    let rows = g.length
    let cols = g[0].length
    let U = 0, L = 0
    let D = rows - 1, R = cols - 1 //, D
    let i: number;
    while (U <= D && L <= R) {
        console.log(U,D,L,R)
        i = L
        while (i <= R) {
            res.push(g[U][i])
            ++i
        }
        U++
        i = U
        while (i <= D) {
            res.push(g[i][R])
            ++i
        }
        R--
        i = R
        if (U <= D) {
            while (i >= L) {
                res.push(g[D][i])
                --i
            }
        }
        --D
        i = D
        if (L <= R) {
            while (i >= U) {
                res.push(g[i][L])
                --i
            }
        }
        ++L
    }
    // console.log(res)
    return res
};

File: leet_0054_spiral_matrix.go
func spiralOrder(m [][]int) []int {
    res := []int {}
    rows := len(m)
    cols := len(m[0])
    up := 0
    down := rows - 1
    left := 0
    right := cols - 1
    var i int
    for up <= down && left <= right {
        i = left
        for i <= right {
            res = append(res, m[up][i])
            i++
        }
        up++

        i = up
        for i <= down {
            res = append(res, m[i][right])
            i++
        }
        right--

        i = right
        if up <= down {
            for i >= left {
                res = append(res, m[down][i])
                i--
            }
        }
        down--

        i = down
        if left <= right {
            for i >= up {
                res = append(res, m[i][left])
                i--
            }
        }
        left++
    }
    return res
}
